File: Prac-2-Aux121/Manejo de Excepciones/Ejercicio4/Inventario.java
import java.util.ArrayList;
import java.util.List;
public class Inventario {
    private List<Producto> productos;
    public Inventario() {
        this.productos = new ArrayList<>();
    }
    public void agregarProductos(Producto producto) {
        for (Producto p : productos) {
            if (p.getCodigo().equalsIgnoreCase(producto.getCodigo())) {
                throw new IllegalArgumentException(producto.getNombre()+"... el codigo del producto ya existe " + p.getCodigo());
            }
        }
            if (producto.getPrecio() < 0) {
                throw new IllegalArgumentException(producto.getNombre()+"... el precio del producto es negativo");
            }
            else {
                if (producto.getStock() < 0) {
                    throw new IllegalArgumentException(producto.getNombre()+"... el stock del producto es negativo");
                }
                else {
                    System.out.println("Se añadio exitosamente el "+producto);
                    productos.add(producto);
                }
            }

    }
    public Producto buscarProducto(String codigo) throws ProductoNoEncontradoException{
        boolean t=false;
        for (Producto p: productos){
            if (p.getCodigo().equalsIgnoreCase(codigo)){
                System.out.println("Se encontro el producto "+p.getNombre()+", su codigo es: "+p.getCodigo());
                t=true;
                Producto prod=new Producto(p.getNombre(),p.getCodigo(),p.getPrecio(),p.getStock());
                return prod;
            }
        }
        if (t!=true){
            throw new ProductoNoEncontradoException("Producto no encontrado con codigo: "+codigo);
        }
        return null;
    }
    public void venderProducto(String codigo, int cantidad) throws StockInsuficienteException, ProductoNoEncontradoException {
        System.out.println("Buscando el producto...");
        Producto p=buscarProducto(codigo);
        if (p.getStock()<cantidad){
            throw new StockInsuficienteException("El stock es insuficiente para vender la cantidad de "+cantidad+" unidades");
        }
        else {
            p.setStock(p.getStock()-cantidad);
            System.out.println("Se vendio exitosamente la cantidad de "+cantidad+" unidades, el nuevo stock es de "+p.getStock()+" unidades");
        }
    }
    public void mostrarInventario(){
        System.out.println("--------------------");
        System.out.println("El inventario actual");
        System.out.println("--------------------");
        for (Producto p: productos){
            System.out.println(p);
        }
    }
  }

File: Prac-2-Aux121/Manejo de Excepciones/Ejercicio4/Inventario.py
class ProductoNoEncontradoException(Exception):
    pass

class StockInsuficienteException(Exception):
    pass

class Producto:
    def __init__(self, codigo, nombre, precio, stock):
        self.codigo = codigo
        self.nombre = nombre
        self.precio = precio
        self.stock = stock

    def __str__(self):
        return f"Producto(codigo={self.codigo}, nombre={self.nombre}, precio={self.precio}, stock={self.stock})"

class Inventario:
    def __init__(self):
        self.productos = []

    def agregarProducto(self, p: Producto):
        if any(prod.codigo == p.codigo for prod in self.productos):
            raise Exception(f"El código {p.codigo} ya existe.")
        if p.precio < 0:
            raise Exception("El precio no puede ser negativo.")
        if p.stock < 0:
            raise Exception("El stock no puede ser negativo.")
        self.productos.append(p)

    def buscarProducto(self, codigo: str) -> Producto:
        for prod in self.productos:
            if prod.codigo == codigo:
                return prod
        raise ProductoNoEncontradoException(f"Producto con código {codigo} no encontrado.")

    def venderProducto(self, codigo: str, cantidad: int):
        producto = self.buscarProducto(codigo)
        if producto.stock >= cantidad:
            producto.stock -= cantidad
            print(f"Venta realizada. Stock restante de {producto.nombre}: {producto.stock}")
        else:
            raise StockInsuficienteException(f"Stock insuficiente para vender {cantidad} unidades de {producto.nombre}.")

if __name__ == "__main__":
    inventario = Inventario()

    try:
        p1 = Producto("001", "Manzana", 0.5, 100)
        inventario.agregarProducto(p1)
        p2 = Producto("002", "Banana", 0.3, 50)
        inventario.agregarProducto(p2)

        inventario.venderProducto("001", 20) 
        inventario.venderProducto("002", 60) 
    except Exception as e:
        print(f"Error: {e}")